File: probes/registry-quickstarts/audit.mjs
import assert from "node:assert/strict";
import { readFileSync, realpathSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { resolve, sep } from "node:path";
import { canonicalStringify, computeStructureHash, validateDocument } from "@agent-topology/spec";

const root = realpathSync("node_modules") + sep;
const imports = {};
const versions = {};
for (const name of ["@agent-topology/spec", "@agent-topology/langgraph", "@langchain/langgraph"]) {
  const path = realpathSync(fileURLToPath(import.meta.resolve(name)));
  assert.ok(path.startsWith(root), "import escaped isolated node_modules");
  imports[name] = path;
  versions[name] = JSON.parse(readFileSync(resolve("node_modules", name, "package.json"))).version;
}
assert.equal(versions["@agent-topology/spec"], "0.1.0-beta.2");
assert.equal(versions["@agent-topology/langgraph"], "0.1.0-beta.2");
assert.equal(versions["@langchain/langgraph"], "1.4.14");
const checks = process.argv.slice(2).map((file) => {
  const raw = readFileSync(file, "utf8");
  const doc = JSON.parse(raw);
  assert.equal(validateDocument(doc).valid, true);
  assert.deepEqual(computeStructureHash(doc), doc.structureHash);
  assert.equal(canonicalStringify(doc) + "\n", raw);
  return {file: file.split("/").at(-1), valid: true, hash: doc.structureHash, canonical: true};
});
console.log(JSON.stringify({runtime: process.version, mode: "ESM audit; examples .mjs and .cjs with dynamic import", imports, versions, checks}, null, 2));

File: probes/registry-quickstarts/audit.py
"""Separate installed-package observation; does not import or modify graph.py."""
import importlib
import importlib.metadata as metadata
import json
from pathlib import Path
import sys

from agent_topology.spec import canonical_json, compute_structure_hash, validate_document


def require(value, message):
    if not value:
        raise ValueError(message)


prefix = Path(sys.prefix).resolve()
versions = {name: metadata.version(name) for name in
            ("agent-topology-spec", "agent-topology-langgraph", "langgraph")}
require(versions["agent-topology-spec"] == versions["agent-topology-langgraph"] == "0.1.0b2",
        "wrong published topology versions")
imports = {}
for name in ("agent_topology.spec", "agent_topology.langgraph", "langgraph.graph"):
    path = Path(importlib.import_module(name).__file__).resolve()
    require(path.is_relative_to(prefix), "import escaped venv")
    imports[name] = str(path)
for dist in metadata.distributions():
    require(Path(dist.locate_file("")).resolve().is_relative_to(prefix), "distribution escaped venv")
    require(dist.read_text("direct_url.json") is None, "direct/editable distribution")
checks = []
for filename in sys.argv[1:]:
    raw = Path(filename).read_text()
    doc = json.loads(raw)
    require(validate_document(doc) == [], "invalid topology")
    require(compute_structure_hash(doc) == doc["structureHash"], "hash mismatch")
    require(canonical_json(doc) + "\n" == raw, "noncanonical serialization")
    checks.append({"file": Path(filename).name, "valid": True, "hash": doc["structureHash"],
                   "canonical": True})
print(json.dumps({"runtime": sys.version, "prefix": str(prefix), "imports": imports,
                  "versions": versions, "sysPath": sys.path, "checks": checks}, indent=2))
